### ISSUE TITLE
Enable string cache by default

### ISSUE THREAD

**Issue by mcrumiller** [pola-rs/polars#9106]
### Problem description

Categorical variables are used to represent strings with numbers for many performance reasons. However, new users are *constantly* tripped up by the string cache requirement when using categoricals from multiple frames. The global string cache is currently disabled by default, meaning that categorical variables from two frames are not comparable by default.

I can't imagine many use cases in which the string cache should be disabled. If a user performs an operation in which two frames with the same categorical strings interact, it his highly unlikely that they intend for those strings not to evaluate equal. I use categoricals all over the place and I have *never* operated with the string cache disabled, in fact it's become almost ritual to add `pl.enable_string_cache(True)` as a prelude to all of my python modules and scripts.

I personally think we should go even one step further and remove the string cache concept entirely (and have all categoricals equate by value), but I realize that is an extreme position here and may result in difficulty with lexical vs physical. What do others think?

**Comment by ghuls** [comment]
It would be nice to have a function to make to local string caches compatible.

I have some library code where users can read different files (in separate functions) where some columns get converted to categoricals. Later those dataframes can be used in other functions. Now I have to enable the global string cache in the library and hope users don't turn it off in between calls.

Having a `with pl.UnifyCat:` context manager to unify local string caches would help a lot.

Remapping local string caches to an unified local string cache should not be too expensive after the unified categorical mapping is created as you can use indexing to lookup the remapped indices values.

```
Local string cache1:
0 : "B"
1: "k"
2: "Aa"
3: "A"
4: "e"

local_string_cache1 = np.array(["B", "k", "Aa", "A", "e"])

Local string cache: cat2:
0 : "f"
1: "Aa"
2: "C"
3: "A"
4: "g"

local_string_cache2 = np.array(["f", "Aa", "C", "A", "g"])

Unified string cache:
0 : "B"
1: "k"
2: "Aa"
3: "A"
4: "e"
5: "f"
6: "C"
7: "g"


unified_string_cache = np.array(["B", "k", "Aa", "A", "e", "f", "C", "g"])

Then to fast map the second cat2 to the unified string cache encoding create an array of length of max integer in cat2.
Then for each index position in this array fill in the corresponding int number of the unified string cache.

array index --> unified string cache encoding
0: 5 ("f")
1: 2 ("Aa")
2: 6 ("C")
3: 3 ("A")
4: 7 ("g")

local_string_cache2_to_unified_string_cache_mapping = np.array([5, 2, 6, 7, 3])

cat1 = [0, 1, 2, 3, 4, 2, 1, 4, 0]
cat2 = [0, 1, 2, 3, 4, 2, 1, 4, 0]

```

```python
import numpy as np

# Local string cache1: int to str mapping.
local_string_cache1 = np.array(["B", "k", "Aa", "A", "e"])

# Local string cache2: int to str mapping.
local_string_cache2 = np.array(["f", "Aa", "C", "A", "g"])

# Unified string cache: int to str mapping.
unified_string_cache = np.array(["B", "k", "Aa", "A", "e", "f", "C", "g"])

# Map local string cache2 int to correct int in unified string cache.
local_string_cache2_to_unified_string_cache_mapping = np.array([5, 2, 6, 7, 3])

# 2 categoricals with local string cache 1 and 2 respectively. 
cat1 = [0, 1, 2, 3, 4, 2, 1, 4, 0]
cat2 = [0, 1, 2, 3, 4, 2, 1, 4, 0]

# Get string representation for categorical 1 via local string cache 1.
In [11]: local_string_cache1[cat1]
Out[11]: array(['B', 'k', 'Aa', 'A', 'e', 'Aa', 'k', 'e', 'B'], dtype='<U2')

# Get string representation for categorical 1 via unified string cache.
In [14]: unified_string_cache[cat1]
Out[14]: array(['B', 'k', 'Aa', 'A', 'e', 'Aa', 'k', 'e', 'B'], dtype='<U2')


# Get string representation for categorical 2 via local string cache 2.
In [12]: local_string_cache2[cat2]
Out[12]: array(['f', 'Aa', 'C', 'A', 'g', 'C', 'Aa', 'g', 'f'], dtype='<U2')

# Get string representation for categorical 2 by first mapping cat2 integers to unified string cache integers and then looking them up in the unified string cache.
In [13]: unified_string_cache[local_string_cache2_to_unified_string_cache_mapping[cat2]]
Out[13]: array(['f', 'Aa', 'C', 'g', 'A', 'C', 'Aa', 'A', 'f'], dtype='<U2')
```

**Comment by magarick** [comment]
> This would keep all the string data in memory, and potentially keep huge strings alive. I'd rather have users to be aware if we do something potentially expensive.
> 
> And with the context manager, they only keep the string cache alive for the duration it needs to be.

For this to be a problem, you'd have to have many large strings that exist transiently and aren't repeated often. Is there a specific case/problem you're thinking of? In my experience, you're much more likely to have a smaller number of strings that get repeated a lot. R and Python both cache strings and I've never heard of this causing a problem. I'd bet it's almost always _less_ expensive to have the cache so I think having to turn it off is the special case where users need to be aware.

Maybe it's possible to split the difference and warn if the cache gets large or have some operations trigger a check to see if anything can be culled.

This also sounds like it's related to the issue's I'm having with https://github.com/pola-rs/polars/pull/9580

**Comment by magarick** [comment]
It seems like we're thinking about this from different perspectives and as a result talking past each other. No matter what there's going to be a tradeoff. My contention is that for a typical user who doesn't know or care about the internals that the current state of affairs might be more difficult for them.

> > In my experience, you're much more likely to have a smaller number of strings that get repeated a lot. 
> 
> Sure, but I have encountered extremas a lot. People use millions of unique strings and very long strings as well. 

Let's say you have on the order of 100k strings of length 1000 and 10M of length 10. You're talking a few hundred MB. As soon as you start duplicating these, a cache helps. That being said maybe work is needed to ensure we're not storing everything forever and warn users if it's going to be a problem.

> If you want to enable the string cache on start up. You can, set the config on startup.

As the project grows I don't imagine this is something most users will want to fiddle with, and rightly so. The question is which tradeoffs have to be made. We're scientists and engineers here so we have to deal in data. The right default depends on the relative frequencies of different use cases and how severe the problems for each are. I don't know these numbers but I'm more than happy to help figure them out.

> > I'd bet it's almost always less expensive to have the cache so I think having to turn it off is the special case where users need to be aware.
> 
> Construction of categorical without a string cache has zero contention. This will be much faster if you convert many columns in parallel. 
> 
But it seems like it leads to trouble with joins and possibly my issue where creating a categorical in an over expression is troublesome. My view is that making categoricals easy to work with and as transparent to the user as possible is important. So if there are other ways to do that I don't care about the cache and I'm happy to work on them with you.

**Comment by ritchie46** [comment]
I am opposed to silently have cost in polars. That cost should be opt-in. That's why we have this string cache as opt in and we don't reallocate on joins.

For scientists who don't care, they can turn on the string cache on startup. If we haven't documented that we can add that in a place more prevalent such as the user guide? 

I understand that can make things a bit more complicated, and I am open to ideas, but I don't want the string cache active as default as it is a hidden cost.

>  That being said maybe work is needed to ensure we're not storing everything forever and warn users if it's going to be a problem.

There is also complexity involved if we start figuring out the lifetime of string caches. How do we know when to clean and how can we be sure that a later construction doesn't need an earlier version of the cache.

I want the lifetime/cache duration to be a conscious decision.

**Comment by jonashaag** [comment]
FWIW in the use cases I deal with on a daily basis we have more like order of multiple 100 M rows with strings either length 10 or 100, sometimes 1000 (although at that point we typically have some kind of chunking so not 100 M rows)

**Comment by mcrumiller** [comment]
@ritchie46 I agree with the sentiment, the issue is the performance case occurs (I think) in a tiny minority of cases.

Why not issue a performance warning when the string cache gets large, with hint to disable it? Similar to panda's fragmented dataframe warning when too many columns are inserted without a rechunk.

**Comment by MarcoGorelli** [comment]
May I suggest another solution?

That the error message include a suggestion to use `enable_string_cache` if the performance cost is acceptable

Kind of like how Rust may tell you
```
help: consider cloning the value if the performance cost is acceptable
```

This preserves "cost being opt-in", whilst giving a quick solution to "scientists who don't care"

**Comment by ritchie46** [comment]
> That the error message include a suggestion to use enable_string_cache if the performance cost is acceptable

This. 

I know we already do that on joins. We could elaborate the error message with:

1. A snippet that sets a global string cache
2. A snippet that sets a local one with a context manager.

It is a single config line opt-in, if you don't care. We could take your hand in the error message.  

I want to follow rust here. Don't hide the complexity, but give informative errors and fail fast.

As a new feature we can add a static analysis on the query that raises an error if we know you haven't set a stringcache on comparisons before we run the query.

**Comment by MarcoGorelli** [comment]
thanks - is it OK for snippets to be written using the Python API? That's probably what most users use, and others can figure out how it maps to their language

**Comment by magarick** [comment]
> I am opposed to silently have cost in polars. That cost should be opt-in. That's why we have this string cache as opt in and we don't reallocate on joins.

Unfortunately, you have a silent cost either way. The only difference is what situations cause the cost and how big it is. If you have a lot of strings AND few repetitions AND the strings are transient, yes you'll waste memory if you don't prune your cache. But consider another common case. You have many repetitions spread out across multiple categorical columns, OR you have few repetitions per column, but multiple similar columns which can easily happen if you copy a column to modify it slightly and keep the original. The threshold for actually using _less_ memory is low here.
As some say, what you have isn't a problem, it's a dilemma.

> For scientists who don't care, they can turn on the string cache on startup. If we haven't documented that we can add that in a place more prevalent such as the user guide?

Isn't this the group of people whom you expect to know the least and want to make it easiest on? If someone's and engineer who needs to know and cares about this stuff, they can change the settings. But I'm looking at this from the perspective of someone who wants things to "just work". 

> I understand that can make things a bit more complicated, and I am open to ideas, but I don't want the string cache active as default as it is a hidden cost.

As established above, it's not always a cost. It comes down to likely use cases and the costs incurred in each. Individual stories are helpful but don't get us to an answer of what the correct, probability-informed default is. I've never seen string caches described as being a cost or a penalty. Quite the opposite. Which is why I'm assuming that on average not having a global cache might cost (a lot) more.

> > That being said maybe work is needed to ensure we're not storing everything forever and warn users if it's going to be a problem.
> 
> There is also complexity involved if we start figuring out the lifetime of string caches. How do we know when to clean and how can we be sure that a later construction doesn't need an earlier version of the cache.

I realize this may be complicated and out of scope, at least for the time being, but is there any way to keep track of when categorical series are created and destroyed? Obviously you don't want to build a full-blown custom garbage collector, but maybe it's possible at some point to clean up the cache periodically when we know some strings aren't being used anymore. Apparently R does something like this.

> I want the lifetime/cache duration to be a conscious decision.
Opt-in or opt-out, it still is, no? You just need to alert people to different conditions.

**Comment by magarick** [comment]
Thinking some more, it seems we may be conflating two different ideas.

To some, including me, a categorical (or factor in R lingo) represents a relatively small number of meaningful discrete values, generally for defining groups. And there are actually two notions of equality. The obvious one is just comparing the label mappings, but the other, which kind of corresponds to no global string cache is that two factors/categoricals aren't comparable if the set of possible levels differs, even if some levels aren't represented in the data. Regardless, it's always something you want to be able to compare and combine.

On the other hand, if you have millions of non-repeated levels that are potentially not sticking around, there's a good chance it's not for defining groups or to have the levels encoded to be used in a regression. Often you'll extract features from the strings, filter/group by their _characteristics_ and possibly drop the original values. These really are strings, not categoricals. But judging from the documentation here (https://github.com/pola-rs/polars/blob/main/polars/src/docs/performance.rs) making a string array categorical makes some operations like reordering values easier.

In both cases there are reasons to not have one big `Utf8Chunked` but not the same reasons. Maybe we're dealing with two very similar but not identical types. After all, even R made `stringsAsFactors` false by default a few years ago. If you make a categorical, you do so intentionally with a specific use in mind and I'd imagine it's the former rather than the latter. So, is there a way to have two "Categorical" types where one uses a global cache to make them behave in the natural way for meaningful groups of the data and another that uses the underlying machinery to make working with strings easier?

**Comment by ritchie46** [comment]
> Isn't this the group of people whom you expect to know the least and want to make it easiest on?

No, I want to make a good tool with good decisions from an engineering stand point. I think it is critical to learn about the tool you are using and understand how to use it. If you set a global string cache, it will have implications and I hope people will understand them.

The PR for the improved error messages is already there: #9715. This should make things easy enough for scientist. The error message clearly states what you can do to fix the error.


Anyone reading this, import polars like this if you don't care or you want a global string cache for the categorical as factors behavior as mentioned by @magarick :

```python
import polars as pl
pl.enable_string_cache(True)
```

I think we should stop discussion further, as there isn't an insurmountable problem. It is a single line fix.

**Comment by jonashaag** [comment]
I benchmarked how costly it is to always run a HyperLogLog based `approx_unique` when constructing utf8/binary series. It adds around 2% of cost to constructing a series in my benchmarks. Maybe this is an acceptable cost to be able to warn about the lack of use of string cache automatically?

**Comment by ritchie46** [comment]
This is not a situation under contention. The global string cache is behind a mutex and will kill parallelism.

I will close this as not planned. It is not something I want in by default and but can easily be opted in as stated in my earlier comment

**Comment by jonashaag** [comment]
I don't understand what you are referring to. I wonder if you didn't properly read my comment, or you didn't refer to it in the first place.

**Comment by ritchie46** [comment]
> but can easily be opted in as stated in my earlier comment

This refers to https://github.com/pola-rs/polars/issues/9106#issuecomment-1621178844

> I benchmarked how costly it is to always run a HyperLogLog based approx_unique when constructing utf8/binary series. It adds around 2% of cost to constructing a series in my benchmarks. 

I referred to this when saying that doing a benchmark on a single column isn't a loadcase under contention. E.g. you don't create categoricals from 100s of columns at a time.

**Comment by jonashaag** [comment]
Ok but what does a HLL based unique cardinality estimation  have to do with the string cache? It doesn't use the string cache and I don't see how it needs mutexes

**Comment by ritchie46** [comment]
>  It doesn't use the string cache and I don't see how it needs mutexes

I don't follow? Why are we discussing that in this thread then? :sweat_smile:
